prov/mrail/src/mrail_rma.c

```diff
@@ -0,0 +1,428 @@
+			}
+		}
+
+		if (ret != 0) {
+			if (ret == -FI_EAGAIN) {
+				break;
+			}
+			/* TODO: Handle errors besides FI_EAGAIN */
+			assert(0);
+		}
+		req->pending_subreq--;
+	}
+
+	return ret;
+}
+
+static struct mrail_req *mrail_dequeue_deferred_req(struct mrail_ep *mrail_ep)
+{
+	struct slist_entry *e;
+
+	e = slist_remove_head(&mrail_ep->deferred_reqs);
+	if (!e) {
+		return NULL;
+	}
+	return container_of(e, struct mrail_req, entry);
+}
+
+static inline void mrail_push_deferred_req(struct mrail_ep *mrail_ep,
+		struct mrail_req *req)
+{
+	slist_insert_head(&req->entry, &mrail_ep->deferred_reqs);
+}
+
+static inline void mrail_enqueue_deferred_req(struct mrail_ep *mrail_ep,
+		struct mrail_req *req)
+{
+	slist_insert_tail(&req->entry, &mrail_ep->deferred_reqs);
+}
+
+void mrail_progress_deferred_reqs(struct mrail_ep *mrail_ep)
+{
+	struct mrail_req *req;
+	ssize_t ret;
+
+	req = mrail_dequeue_deferred_req(mrail_ep);
+	while (req) {
+		ret = mrail_try_posting_req(req);
+		if (ret) {
+			mrail_push_deferred_req(mrail_ep, req);
+			break;
+		}
+		req = mrail_dequeue_deferred_req(mrail_ep);
+	}
+}
+
+static size_t mrail_prepare_rma_subreqs(struct mrail_ep *mrail_ep,
+		const struct fi_msg_rma *msg, struct mrail_req *req)
+{
+	ssize_t ret;
+	struct mrail_subreq *subreq;
+	size_t subreq_count;
+	size_t total_len;
+	size_t chunk_len;
+	size_t subreq_len;
+	size_t iov_index;
+	size_t iov_offset;
+	size_t rma_iov_index;
+	size_t rma_iov_offset;
+	int i;
+
+	/* For now, stripe across all rails.
+	 * This could be determined by a dynamic scheduler instead.
+	 */
+	subreq_count = mrail_ep->num_eps;
+
+	total_len = ofi_total_iov_len(msg->msg_iov, msg->iov_count); 
+	chunk_len = total_len / subreq_count;
+
+	/* The first chunk is the longest */
+	subreq_len =  chunk_len + (total_len % subreq_count);
+	iov_index = 0;
+	iov_offset = 0;
+	rma_iov_index = 0;
+	rma_iov_offset = 0;
+
+	/* The array is filled in reverse order -- i.e. first subreq at
+	 * last position in the array. Filling the array in this order saves
+	 * us from having to use two variables, to track the total number of
+	 * subreqs, and to know which one to try posting next.
+	 * Instead, a single variable (req->pending_subreq) is used to keep
+	 * track of which subreq to post next, starting at the end of the
+	 * array.
+	 */
+	for (i = (subreq_count - 1); i >= 0; --i) {
+		subreq = &req->subreqs[i];
+
+		subreq->parent = req;
+
+		ret = ofi_copy_iov_desc(subreq->iov, subreq->descs,
+				&subreq->iov_count,
+				(struct iovec *)msg->msg_iov, msg->desc,
+				msg->iov_count, &iov_index, &iov_offset,
+				subreq_len);
+		if (ret) {
+			return 0;
+		}
+
+		ret = ofi_copy_rma_iov(subreq->rma_iov, &subreq->rma_iov_count,
+				(struct fi_rma_iov *)msg->rma_iov,
+				msg->rma_iov_count, &rma_iov_index,
+				&rma_iov_offset, subreq_len);
+		if (ret) {
+			return 0;
+		}
+
+		/* All the other chunks have the same length */
+		subreq_len = chunk_len;
+	}
+
+	return subreq_count;
+}
```

Comment:
This function return value differs from most of code, in that success is non-zero and error is 0.  It would be better to pass failure codes through (e.g. from ofi_copy_iov_desc), so that they don't get lost.